presto-orc/src/main/java/com/facebook/presto/orc/reader/BooleanSelectiveStreamReader.java

```diff
@@ -0,0 +1,332 @@
+        dataStreamSource = missingStreamSource(BooleanInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public void startRowGroup(InputStreamSources dataStreamSources)
+    {
+        presentStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, PRESENT, BooleanInputStream.class);
+        dataStreamSource = dataStreamSources.getInputStreamSource(streamDescriptor, DATA, BooleanInputStream.class);
+
+        readOffset = 0;
+
+        presentStream = null;
+        dataStream = null;
+
+        rowGroupOpen = false;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .addValue(streamDescriptor)
+                .toString();
+    }
+
+    @Override
+    public void close()
+    {
+        systemMemoryContext.close();
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE + sizeOf(values) + sizeOf(nulls) + sizeOf(outputPositions);
+    }
+
+    private void openRowGroup()
+            throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+        dataStream = dataStreamSource.openStream();
+        rowGroupOpen = true;
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        boolean nullsAllowed = filter == null || filter.testNull();
+        boolean recordNulls = nullsAllowed && presentStream != null;
+        if (outputRequired) {
+            ensureValuesCapacity(positionCount, recordNulls);
+        }
+
+        if (filter != null) {
+            ensureOutputPositionsCapacity(positionCount);
+        }
+        else {
+            outputPositions = positions;
+        }
+
+        if (readOffset < offset) {
+            skip(offset - readOffset);
+        }
+
+        int streamPosition = 0;
+        allNulls = false;
+        if (dataStream == null && presentStream != null) {
+            // all nulls
+            presentStream.skip(positions[positionCount - 1]);
+            streamPosition = positions[positionCount - 1] + 1;
+
+            if (nullsAllowed) {
+                outputPositionCount = positionCount;
+                allNulls = true;
+            }
+            else {
+                outputPositionCount = 0;
+            }
+        }
+        else if (filter == null) {
+            for (int i = 0; i < positionCount; i++) {
+                int position = positions[i];
+                if (position > streamPosition) {
+                    skip(position - streamPosition);
+                    streamPosition = position;
+                }
+
+                if (presentStream != null && !presentStream.nextBit()) {
+                    nulls[i] = true;
```

Comment:
Do we need to check `outputRequired` and `recordNulls` before doing the assignment?

Reply:
- here we have `filter == null`, hence `outputRequired == true`
- we also have a null value which means we have to record; sometimes we don't need to record non-null values, but we always record nulls (except when all values are null, but that case is processed separately)
- would it help to add some comments throughout this code?